File: frontend/lib/skills.ts
/**
 * lib/skills.ts — Single source of truth for all skill metadata.
 *
 * To add a new skill:
 *   1. Add it to SKILL_CATEGORIES under the appropriate confidence bucket
 *   2. Add its label to SKILL_LABELS
 *   3. Add its abbreviation to SKILL_ABBREV
 *   4. Add its type priority to SKILL_TYPE_PRIORITY
 *   5. Add it to the appropriate SKILL_GROUPS entry (legends page)
 *   6. Mirror the change in backend/services/claude_assessment.py (ALL_SKILLS in legends.py and compositing.py derive from it automatically)
 */

// ---------------------------------------------------------------------------
// Canonical skill list grouped by stat confidence
// ---------------------------------------------------------------------------

export const SKILL_CATEGORIES: Record<string, string[]> = {
  "High Confidence": [
    "spot_up_shooter",
    "off_dribble_shooter",
    "offensive_rebounder",
    "rebounder",
    "rim_protector",
    "isolation_scorer",
  ],
  Moderate: [
    "movement_shooter",
    "cutter",
    "transition_threat",
    "pnr_ball_handler",
    "pnr_finisher",
    "crafty_finisher",
    "driver",
    "vertical_spacer",
    "screen_setter",
    "passer",
    "mid_post_player",
    "low_post_player",
  ],
  "Low Confidence": [
    "versatile_defender",
    "perimeter_disruptor",
    "high_flyer",
  ],
};

/** Flat ordered list of all skill names, in display order. */
export const ALL_SKILL_NAMES: string[] = Object.values(SKILL_CATEGORIES).flat();

/** Total number of skills — derive this rather than hardcoding. */
export const TOTAL_SKILLS = ALL_SKILL_NAMES.length;

// ---------------------------------------------------------------------------
// Human-readable display names
// ---------------------------------------------------------------------------

export const SKILL_LABELS: Record<string, string> = {
  spot_up_shooter:          "Spot Up Shooter",
  off_dribble_shooter:      "Off-Dribble Shooter",
  offensive_rebounder:      "Offensive Rebounder",
  rebounder:                "Rebounder",
  rim_protector:            "Rim Protector",
  isolation_scorer:         "Isolation Scorer",
  movement_shooter:         "Movement Shooter",
  cutter:                   "Cutter",
  transition_threat:        "Transition Threat",
  pnr_ball_handler:         "PnR Ball Handler",
  pnr_finisher:             "PnR Finisher",
  crafty_finisher:          "Crafty Finisher",
  driver:                   "Driver",
  vertical_spacer:          "Vertical Spacer",
  screen_setter:            "Screen Setter",
  passer:                   "Passer",
  mid_post_player:          "Mid-Post Player",
  low_post_player:          "Low-Post Player",
  versatile_defender:       "Versatile Defender",
  perimeter_disruptor:      "Perimeter Disruptor",
  high_flyer:               "High Flyer",
};

/** Converts a skill key to its display name, with snake_case title-case fallback. */
export function formatSkillName(name: string): string {
  return (
    SKILL_LABELS[name] ??
    name.split("_").map((w) => w.charAt(0).toUpperCase() + w.slice(1)).join(" ")
  );
}

// ---------------------------------------------------------------------------
// Table column abbreviations (PlayerTable)
// ---------------------------------------------------------------------------

export const SKILL_ABBREV: Record<string, string> = {
  spot_up_shooter:          "Spot Up",
  off_dribble_shooter:      "Off Drib",
  offensive_rebounder:      "Off Reb",
  rebounder:                "Reb",
  rim_protector:            "Rim Prot",
  isolation_scorer:         "Iso",
  movement_shooter:         "Move Shoot",
  cutter:                   "Cutter",
  transition_threat:        "Trans",
  pnr_ball_handler:         "PnR BH",
  pnr_finisher:             "PnR Fin",
  crafty_finisher:          "Crafty",
  driver:                   "Driver",
  vertical_spacer:          "V-Space",
  screen_setter:            "Screener",
  passer:                   "Passer",
  mid_post_player:          "Mid Post",
  low_post_player:          "Lo Post",
  versatile_defender:       "Versa Def",
  perimeter_disruptor:      "Perim Disr",
  high_flyer:               "Hi Fly",
};

// ---------------------------------------------------------------------------
// Card display priority (PlayerCard — lower = shown first)
// ---------------------------------------------------------------------------

export const SKILL_TYPE_PRIORITY: Record<string, number> = {
  // Additive (0) — always valuable to have multiples on a team
  spot_up_shooter:          0,
  movement_shooter:         0,
  rebounder:                0,
  offensive_rebounder:      0,
  rim_protector:            0,
  vertical_spacer:          0,
  screen_setter:            0,
  versatile_defender:      0,
  cutter:                   0,
  // Threshold (1) — valuable when at least one player excels
  off_dribble_shooter:      1,
  crafty_finisher:          1,
  driver:                   1,
  pnr_finisher:             1,
  passer:                   1,
  mid_post_player:          1,
  low_post_player:          1,
  high_flyer:               1,
  perimeter_disruptor:      1,
  // Zero-sum (2) — team typically needs just one at a high level
  isolation_scorer:         2,
  pnr_ball_handler:         2,
  transition_threat:        2,
};

// ---------------------------------------------------------------------------
// Public player profile — 7 horizontal category columns
// Skills within each column render in the order listed here;
// the profile component sorts them highest tier first at render time.
// ---------------------------------------------------------------------------

export const PUBLIC_SKILL_CATEGORIES: Record<string, string[]> = {
  "Perimeter Scoring":     ["spot_up_shooter", "movement_shooter", "off_dribble_shooter"],
  "On-Ball Creation":      ["isolation_scorer", "pnr_ball_handler", "driver"],
  "Off-Ball & Transition": ["cutter", "transition_threat", "pnr_finisher"],
  "Interior Scoring":      ["crafty_finisher", "high_flyer", "vertical_spacer", "mid_post_player", "low_post_player"],
  "Playmaking":            ["passer"],
  "Physicality":           ["screen_setter", "offensive_rebounder", "rebounder"],
  "Defense":               ["rim_protector", "perimeter_disruptor", "versatile_defender"],
};

/** Profile-ordered flat list of all skill names — matches PUBLIC_SKILL_CATEGORIES order. */
export const PROFILE_SKILL_ORDER: string[] = Object.values(PUBLIC_SKILL_CATEGORIES).flat();

// ---------------------------------------------------------------------------
// Legends page grouping (by skill_category type, not confidence)
// ---------------------------------------------------------------------------

export const SKILL_GROUPS: { label: string; skills: string[] }[] = [
  {
    label: "Additive Skills",
    skills: [
      "spot_up_shooter",
      "off_dribble_shooter",
      "isolation_scorer",
      "movement_shooter",
      "cutter",
      "transition_threat",
      "pnr_ball_handler",
      "pnr_finisher",
      "crafty_finisher",
      "passer",
      "offensive_rebounder",
      "vertical_spacer",
    ],
  },
  {
    label: "Threshold-Based Skills",
    skills: [
      "rebounder",
      "rim_protector",
      "screen_setter",
      "driver",
      "mid_post_player",
      "low_post_player",
    ],
  },
  {
    label: "Zero-Sum Skills",
    skills: [
      "versatile_defender",
      "perimeter_disruptor",
      "high_flyer",
    ],
  },
];

File: backend/services/skills.py
"""
skills.py — HISTORICAL REFERENCE / BOOTSTRAP ONLY for the skill taxonomy.

Runtime taxonomy now comes from the active Evaluation Version row in the
database (payload.taxonomy.skills). This file seeds the initial cohesion-v1
Version via backend/scripts/dump_v1_blob.py.

The pipeline (stat evaluation, Claude assessment, compositing) still reads
these constants for Skill Profile creation. The Evaluation Version controls
only cohesion scoring math, not the pipeline.

Confidence tiers determine how skills are rated:
  HIGH      — stat pipeline is reliable; Claude is NOT called
  MODERATE  — Claude runs blind (sees stats but NOT the stat tier)
  LOW       — Claude runs informed (sees stats AND stat tier + confidence)
"""

# ---------------------------------------------------------------------------
# Confidence classification
# ---------------------------------------------------------------------------

# Skills where stat pipeline is reliable — Claude is NOT called
HIGH_CONFIDENCE_SKILLS: frozenset[str] = frozenset({
    "rim_protector",
    "spot_up_shooter",
    "off_dribble_shooter",
    "rebounder",
    "offensive_rebounder",
    "isolation_scorer",
    "steady_hand",
})

# Skills where Claude runs blind (sees stats but NOT the stat tier)
MODERATE_CONFIDENCE_SKILLS: frozenset[str] = frozenset({
    "cutter",
    "movement_shooter",
    "passer",
    "crafty_finisher",
    "driver",
    "mid_post_player",
    "low_post_player",
    "screen_setter",
    "vertical_spacer",
    "transition_threat",
    "pnr_ball_handler",
    "pnr_finisher",
})

# Skills where Claude runs informed (sees stats AND stat tier + confidence)
LOW_CONFIDENCE_SKILLS: frozenset[str] = frozenset({
    "versatile_defender",
    "point_of_attack_defender",
    "off_ball_disruptor",
    "high_flyer",
})

# Skills a bulk "Trust Stats" must never resolve across many players at once.
# Their tier drives an archetype label, and the stats engine is weakest exactly
# there, so a human decides each one.
# #152 split perimeter_disruptor into point_of_attack_defender + off_ball_disruptor; D19 labels read these tiers
NO_BULK_TRUST_STATS_SKILLS: frozenset[str] = frozenset({
    "versatile_defender",
    "point_of_attack_defender",
    "off_ball_disruptor",
})

# Composite entries whose `source` records a human review decision (issue #120).
# A recompute must never silently overwrite these — resolved (a flag adjudicated
# in /admin/review) and manual_override (an admin's direct tier) are people's
# calls, not stat output. One definition, because the recompute path and the
# whole-profile guard must agree on what a person's call looks like.
HUMAN_DECISION_SOURCES: frozenset[str] = frozenset({"resolved", "manual_override"})

# ---------------------------------------------------------------------------
# Derived collections
# ---------------------------------------------------------------------------

# Flat sorted list of every skill key — derived, never hardcoded
ALL_SKILLS: list[str] = sorted(
    HIGH_CONFIDENCE_SKILLS | MODERATE_CONFIDENCE_SKILLS | LOW_CONFIDENCE_SKILLS
)

# ---------------------------------------------------------------------------
# Human-readable definitions (used in Claude prompts)
# ---------------------------------------------------------------------------

SKILL_DEFINITIONS: dict[str, str] = {
    "spot_up_shooter":     "Hits catch-and-shoot three-pointers and mid-range shots from set positions.",
    "off_dribble_shooter": "Creates and converts shots off the dribble, including pull-ups and step-backs.",
    "isolation_scorer":    "Beats defenders one-on-one in isolation situations through dribble moves and athleticism.",
    "movement_shooter":    "Hits shots while relocating off screens and handoffs (not just standing still).",
    "cutter":              "Scores effectively by cutting to the basket off-ball.",
    "transition_threat":   "Scores effectively in the open court on fast breaks.",
    "pnr_ball_handler":    "Initiates and scores/creates effectively as the ball handler in pick-and-roll actions.",
    "pnr_finisher":        "Scores effectively as the screener in pick-and-roll actions, whether rolling, popping, or slipping.",
    "crafty_finisher":     "Scores at the rim using touch, body control, and foul-drawing ability rather than pure athleticism.",
    "driver":              "Consistently attacks the paint from the perimeter off the dribble, generating driving lane pressure and paint touches.",
    "passer":              "Creates quality shot opportunities for teammates through vision and passing skill.",
    "offensive_rebounder": "Consistently crashes offensive boards and converts second-chance opportunities.",
    "vertical_spacer":     "Threatens vertically as a lob target and above-the-rim finisher, creating driving lanes for teammates.",
    "rebounder":           "Consistently secures defensive boards through positioning, boxing out, and effort.",
    "rim_protector":       "Deters and blocks shots at the rim, altering opponent finishing attempts.",
    "screen_setter":       "Sets quality screens that free teammates for open shots.",
    "mid_post_player":     "Scores effectively from the mid-post/elbow area using face-up moves and mid-range shooting.",
    "low_post_player":     "Scores effectively with back-to-basket moves in the low post.",
    "versatile_defender":  "Can guard multiple positional groups effectively when switched.",
    "point_of_attack_defender": "Contains the ball handler, fights over screens, and takes the toughest perimeter assignment when asked.",
    "off_ball_disruptor":  "Disrupts the offense away from his own man: holds the right help position, rotates and recovers on the perimeter, digs at drivers and jumps passing lanes, creating deflections, steals and charges.",
    "high_flyer":          "Possesses elite explosive athleticism for above-the-rim plays, highlight dunks, and transition finishes.",
    "steady_hand":      "Protects possessions with a low turnover rate relative to ball responsibility — secure handling, safe decisions, and strong hands under pressure.",
}

# Display names for every Skill. Mirrors SKILL_LABELS in frontend/lib/skills.ts —
# keep the two in step. The archetype why-lines read this (never
# claude_assessment._SKILL_DISPLAY_NAMES, which covers only the Claude-rated
# Skills and pulls in the anthropic and supabase imports).
SKILL_LABELS: dict[str, str] = {
    "spot_up_shooter":          "Spot Up Shooter",
    "off_dribble_shooter":      "Off-Dribble Shooter",
    "offensive_rebounder":      "Offensive Rebounder",
    "rebounder":                "Defensive Rebounding",
    "rim_protector":            "Rim Protector",
    "isolation_scorer":         "Isolation Scorer",
    "steady_hand":              "Steady Hand",
    "movement_shooter":         "Movement Shooter",
    "cutter":                   "Cutter",
    "transition_threat":        "Transition Threat",
    "pnr_ball_handler":         "PnR Ball Handler",
    "pnr_finisher":             "PnR Finisher",
    "crafty_finisher":          "Below the Rim Finishing",
    "driver":                   "Driver",
    "vertical_spacer":          "Vertical Spacer",
    "screen_setter":            "Screen Setter",
    "passer":                   "Passer",
    "mid_post_player":          "Mid-Post Player",
    "low_post_player":          "Low-Post Player",
    "versatile_defender":       "Versatile Defender",
    "point_of_attack_defender": "Point of Attack Defender",
    "off_ball_disruptor":       "Off-Ball Disruptor",
    "high_flyer":               "Above the Rim Finishing",
}


# ---------------------------------------------------------------------------
# Legacy key alias (#152)
# ---------------------------------------------------------------------------

# The one Skill the #152 split renamed. Old rows — a pre-split Snapshot Release,
# a saved team, an Evaluation Version blob whose formulas still name the old key
# — carry `perimeter_disruptor`; the code now reads `point_of_attack_defender`.
# The alias runs both ways so either side can be the one that is current.
_LEGACY_SKILL_ALIASES: tuple[tuple[str, str], ...] = (
    ("perimeter_disruptor", "point_of_attack_defender"),
)


def with_legacy_skill_keys(skills: dict) -> dict:
    """Return a copy of ``skills`` where each renamed Skill answers to both keys.

    A present ``perimeter_disruptor`` fills an absent ``point_of_attack_defender``
    and the reverse. A key already present is never overwritten, so a post-split
    profile holding both keeps its own on-ball tier.

    Key presence is what the engine reads (``present_keys`` in
    ``composites.compute_raw_composites`` separates "rated None" from "never
    rated"), so the alias copies the value as it is — an entry dict, a bare tier
    string, or None for an unrated Skill.

    # ponytail: delete after prod runs EV v10 and a post-split release
    """
    aliased = dict(skills or {})
    for old, new in _LEGACY_SKILL_ALIASES:
        if old in aliased and new not in aliased:
            aliased[new] = aliased[old]
        elif new in aliased and old not in aliased:
            aliased[old] = aliased[new]
    return aliased
